nginx/lua/webdav_content.lua

```diff
@@ -0,0 +1,204 @@
+            -- e.g. a C function for https://en.wikipedia.org/wiki/Adler-32
+            -- libz has this function
+            -- TODO: coroutine https://www.lua.org/manual/5.1/manual.html#2.11
+            adler_state = adler32_increment(adler_state, buffer) 
+            file:write(buffer)
+        end
+    until not buffer
+    ngx.say("TPC complete")
+    file:close()
+    local adler_value = adler32_finalize(adler_state)
+    ngx.say("Adler32 is " .. adler_value)
+    setxattr(destination_localpath, "user.nginx-webdav.adler32", tostring(adler_value))
+
+    -- this allows the connection to be reused by other requests
+    ok, err = httpc:set_keepalive()
+    if not ok then
+        -- TODO: is this a fatal error?
+        ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
+        ngx.say("failed to set keepalive: ", err)
+        return ngx.exit(ngx.OK)
+    end
+end
+
+local function getxattr(path, key, value)
+  -- FIXME this doesn't work at all since we need to receive the string back
+  -- from a pointer to a void*
+  -- https://luajit.org/ext_ffi_tutorial.html
+  return ffi.C.getxattr(path, key, value, string.len(value), 0);
+end
+
+if ngx.var.request_method == "COPY" then
+    -- The COPY method is supported by ngx_http_dav_module but only for files on the same server.
+    -- We intercept the method here to support third-party push copy.
+    -- TODO: is this the best spot in the request lifecycle to do this?
+    -- https://openresty-reference.readthedocs.io/en/latest/Directives/
+    if not ngx.var.http_source then
+        ngx.status = ngx.HTTP_BAD_REQUEST
+        ngx.say("no source provided")
+        return ngx.exit(ngx.OK)
+    end
+
+    if ngx.var.http_destination then
+        ngx.status = ngx.HTTP_NOT_IMPLEMENTED
+        ngx.say("third-party push copy not implemented")
+        return ngx.exit(ngx.OK)
+    end
+
+    -- TODO: better way to find the local file location?
+    third_party_pull(ngx.var.http_source, "/var/www" .. ngx.var.request_uri)
+
+    -- At this point, the connection will either be safely back in the pool, or closed.
+    return ngx.exit(ngx.HTTP_OK)
+elseif ngx.var.request_method == "GET" then
+  -- TODO handle range requests
+  local fd = uv.fs_open("/var/www" .. ngx.var.request_uri, "r", 644)
+  -- Amount we try to read from the filesystem at a time, most distributed
+  -- filesystems have large "block sizes", so lets try a bigger number than
+  -- usual
+  local buffer_size = 64 * 1024 * 1024
+  repeat
+    -- TODO error handling
+    local buffer = uv.fs_read(fd, buffer_size)
+    ngx.print(buffer)
+  until not buffera
+  return ngx.exit(ngx.HTTP_OK)
+elseif ngx.var.request_method == "PUT" then
+  -- TODO write coalescing, we don't want to send a bunch of <1MB writes to a
+  -- distributed filesystem
+  -- TODO we don't support ranged writes (screws with checksum)
+  local fd, err = uv.fs_open("/var/www" .. ngx.var.request_uri, "w", tonumber(644, 8))
```

Comment:
So I think we can use a combination or openresty locks and uv callbacks to make this truly async:
```lua
  local lock, err = resty_lock:new("put_requests")
  
  local fd = nil
  local elapsed, err = lock:lock(ngx.var.request_uri)
  -- if err, exit (already receiving this file)
  uv.fs_open("/var/www" .. ngx.var.request_uri, "w", tonumber(644, 8), function(err, ifd)
    -- do something if err
    fd = ifd
    lock:unlock()
  end)
  
  local elapsed, err = lock:lock(ngx.var.request_uri)
  -- do next async fs request and unlock in callback
  -- etc.
```
a nice bonus here is that we don't dogpile put requests to the same url